File: apps/tabtin-ios/Tabtin/Features/Conversation/MessageBubble/StreamingTailRevealPolicy.swift
import Foundation

/// 流式尾巴只淡「上一帧之后的后缀」。
/// 重写 / 回退整段当 prefix，不动画；token 快于淡入时只跟最新后缀。
enum StreamingTailRevealPolicy {
    static let incomingFadeDuration: TimeInterval = 0.1
    static let caretBlinkDuration: TimeInterval = 0.53
    static let incomingStartOpacity: Double = 0.4

    struct Reveal: Equatable {
        let prefix: String
        let incoming: String
        let shouldAnimateIncoming: Bool
    }

    static func reveal(previousTail: String, nextTail: String) -> Reveal {
        if nextTail.hasPrefix(previousTail) {
            let incoming = String(nextTail.dropFirst(previousTail.count))
            return Reveal(
                prefix: previousTail,
                incoming: incoming,
                shouldAnimateIncoming: !incoming.isEmpty
            )
        }
        return Reveal(prefix: nextTail, incoming: "", shouldAnimateIncoming: false)
    }
}

File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/features/conversation/StreamingTailRevealPolicy.kt
package com.tabtin.mobile.features.conversation

/**
 * 流式尾巴只淡「上一帧之后的后缀」。
 * 重写 / 回退整段当 prefix，不动画；token 快于淡入时只跟最新后缀。
 */
internal object StreamingTailRevealPolicy {
    const val INCOMING_FADE_MS: Int = 100
    const val CARET_BLINK_MS: Int = 530
    const val INCOMING_START_ALPHA: Float = 0.4f

    data class Reveal(
        val prefix: String,
        val incoming: String,
        val shouldAnimateIncoming: Boolean,
    )

    fun reveal(previousTail: String, nextTail: String): Reveal {
        if (nextTail.startsWith(previousTail)) {
            val incoming = nextTail.substring(previousTail.length)
            return Reveal(
                prefix = previousTail,
                incoming = incoming,
                shouldAnimateIncoming = incoming.isNotEmpty(),
            )
        }
        return Reveal(
            prefix = nextTail,
            incoming = "",
            shouldAnimateIncoming = false,
        )
    }
}
